File: Llama3.CSharp/Tokenizer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Llama3.CSharp;

public class Vocabulary
{
    public string[] Tokens { get; }
    public Dictionary<string, int> TokenToIndex { get; }

    public Vocabulary(string[] tokens)
    {
        Tokens = tokens;
        TokenToIndex = new Dictionary<string, int>(tokens.Length);
        for (int i = 0; i < tokens.Length; i++)
        {
            TokenToIndex[tokens[i]] = i;
        }
    }

    public int? GetIndex(string token) => TokenToIndex.TryGetValue(token, out int index) ? index : null;
}

public class Tokenizer
{
    private readonly Vocabulary _vocab;
    private readonly Dictionary<(int, int), int> _merges;
    private readonly Dictionary<string, int> _specialTokens;
    private readonly uint[] _b2u = new uint[256];
    private readonly Dictionary<uint, byte> _u2b = new();

    public Tokenizer(Vocabulary vocab, string[] mergesRaw, Dictionary<string, int> specialTokens)
    {
        _vocab = vocab;
        _specialTokens = specialTokens;
        _merges = new Dictionary<(int, int), int>(mergesRaw.Length);

        foreach (var line in mergesRaw)
        {
            var parts = line.Split(' ');
            if (parts.Length != 2) continue;
            var id1 = _vocab.GetIndex(parts[0]);
            var id2 = _vocab.GetIndex(parts[1]);
            if (id1 == null || id2 == null) continue;

            var combined = parts[0] + parts[1];
            var cid = _vocab.GetIndex(combined);
            if (cid != null)
            {
                _merges[(id1.Value, id2.Value)] = cid.Value;
            }
        }

        InitByteUnicodeMap();
    }

    private void InitByteUnicodeMap()
    {
        var bs = new List<byte>();
        for (int i = '!'; i <= '~'; i++) bs.Add((byte)i);
        for (int i = 0xA1; i <= 0xAC; i++) bs.Add((byte)i);
        for (int i = 0xAE; i <= 0xFF; i++) bs.Add((byte)i);

        var cs = bs.Select(b => (uint)b).ToList();
        uint n = 0;
        for (int b = 0; b < 256; b++)
        {
            if (!bs.Contains((byte)b))
            {
                bs.Add((byte)b);
                cs.Add(256 + n);
                n++;
            }
        }

        for (int i = 0; i < 256; i++)
        {
            _b2u[bs[i]] = cs[i];
            _u2b[cs[i]] = bs[i];
        }
    }

    public int[] Encode(string text, string[] allowedSpecial)
    {
        foreach (var s in allowedSpecial)
        {
            if (text == s && _specialTokens.TryGetValue(s, out int id))
            {
                return new[] { id };
            }
        }

        var ids = new List<int>();
        foreach (var b in Encoding.UTF8.GetBytes(text))
        {
            var token = GetTokenFromByte(b);
            var id = _vocab.GetIndex(token);
            if (id != null) ids.Add(id.Value);
        }

        while (ids.Count >= 2)
        {
            int? bestPairIdx = null;
            int bestIdx = int.MaxValue;

            for (int i = 0; i < ids.Count - 1; i++)
            {
                if (_merges.TryGetValue((ids[i], ids[i + 1]), out int mergeIdx))
                {
                    if (mergeIdx < bestIdx)
                    {
                        bestIdx = mergeIdx;
                        bestPairIdx = i;
                    }
                }
            }

            if (bestPairIdx != null)
            {
                int i = bestPairIdx.Value;
                int mid = bestIdx;
                ids[i] = mid;
                ids.RemoveAt(i + 1);
            }
            else
            {
                break;
            }
        }

        return ids.ToArray();
    }

    public string Decode(int[] tokens)
    {
        var res = new List<byte>();
        foreach (var t in tokens)
        {
            var s = _vocab.Tokens[t];
            foreach (var cp in GetCodepoints(s))
            {
                if (_u2b.TryGetValue(cp, out byte b))
                {
                    res.Add(b);
                }
                else
                {
                    res.AddRange(Encoding.UTF8.GetBytes(char.ConvertFromUtf32((int)cp)));
                }
            }
        }
        return Encoding.UTF8.GetString(res.ToArray());
    }

    private string GetTokenFromByte(byte b)
    {
        uint code = _b2u[b];
        return char.ConvertFromUtf32((int)code);
    }

    private static IEnumerable<uint> GetCodepoints(string s)
    {
        for (int i = 0; i < s.Length; i++)
        {
            if (char.IsHighSurrogate(s[i]) && i + 1 < s.Length && char.IsLowSurrogate(s[i + 1]))
            {
                yield return (uint)char.ConvertToUtf32(s[i], s[i + 1]);
                i++;
            }
            else
            {
                yield return (uint)s[i];
            }
        }
    }

    public Dictionary<string, int> SpecialTokens => _specialTokens;
}

File: Llama3.Rust/src/tokenizer.rs
use std::collections::HashMap;

pub struct Vocabulary {
    pub tokens: Vec<String>,
    pub token_to_index: HashMap<String, u32>,
}

impl Vocabulary {
    pub fn new(tokens: Vec<String>) -> Self {
        let mut token_to_index = HashMap::with_capacity(tokens.len());
        for (i, token) in tokens.iter().enumerate() {
            token_to_index.insert(token.clone(), i as u32);
        }
        Vocabulary { tokens, token_to_index }
    }

    pub fn get_index(&self, token: &str) -> Option<u32> {
        self.token_to_index.get(token).copied()
    }
}

pub struct Tokenizer {
    pub vocab: Vocabulary,
    pub merges: HashMap<(u32, u32), u32>,
    pub special_tokens: HashMap<String, u32>,
    pub b2u: [u32; 256],
    pub u2b: HashMap<u32, u8>,
}

impl Tokenizer {
    pub fn new(vocab: Vocabulary, merges_raw: Vec<String>, special_tokens: HashMap<String, u32>) -> Self {
        let mut merges = HashMap::with_capacity(merges_raw.len());
        for line in merges_raw {
            let parts: Vec<&str> = line.split(' ').collect();
            if parts.len() != 2 { continue; }
            let id1 = vocab.get_index(parts[0]);
            let id2 = vocab.get_index(parts[1]);
            if let (Some(i1), Some(i2)) = (id1, id2) {
                let combined = format!("{}{}", parts[0], parts[1]);
                if let Some(cid) = vocab.get_index(&combined) {
                    merges.insert((i1, i2), cid);
                }
            }
        }

        let mut b2u = [0u32; 256];
        let mut u2b = HashMap::new();

        let mut bs = Vec::new();
        for i in b'!'..=b'~' { bs.push(i); }
        for i in 0xA1..=0xAC { bs.push(i as u8); }
        for i in 0xAE..=0xFF { bs.push(i as u8); }

        let mut cs: Vec<u32> = bs.iter().map(|&b| b as u32).collect();
        let mut n = 0;
        for b in 0..=255 {
            if !bs.contains(&(b as u8)) {
                bs.push(b as u8);
                cs.push(256 + n);
                n += 1;
            }
        }

        for i in 0..256 {
            b2u[bs[i] as usize] = cs[i];
            u2b.insert(cs[i], bs[i]);
        }

        Tokenizer { vocab, merges, special_tokens, b2u, u2b }
    }

    pub fn encode(&self, text: &str, allowed_special: &[&str]) -> Vec<u32> {
        for &s in allowed_special {
            if text == s {
                if let Some(&id) = self.special_tokens.get(s) {
                    return vec![id];
                }
            }
        }

        let mut ids = Vec::new();
        for &b in text.as_bytes() {
            let token = std::char::from_u32(self.b2u[b as usize]).unwrap().to_string();
            if let Some(id) = self.vocab.get_index(&token) {
                ids.push(id);
            }
        }

        while ids.len() >= 2 {
            let mut best_pair = None;
            let mut best_idx = u32::MAX;

            for i in 0..ids.len() - 1 {
                if let Some(&merge_idx) = self.merges.get(&(ids[i], ids[i+1])) {
                    if merge_idx < best_idx {
                        best_idx = merge_idx;
                        best_pair = Some(i);
                    }
                }
            }

            if let Some(i) = best_pair {
                ids[i] = best_idx;
                ids.remove(i + 1);
            } else {
                break;
            }
        }

        ids
    }

    pub fn decode(&self, tokens: &[u32]) -> String {
        let mut res = Vec::new();
        for &t in tokens {
            let s = &self.vocab.tokens[t as usize];
            for c in s.chars() {
                let cp = c as u32;
                if let Some(&b) = self.u2b.get(&cp) {
                    res.push(b);
                } else {
                    res.extend_from_slice(c.to_string().as_bytes());
                }
            }
        }
        String::from_utf8_lossy(&res).to_string()
    }
}
